File: data_structures/data_structures_cpp/heaps.py
import random

class MinHeap:
    def __init__(self):
        self.data = [] # the tree

    def _parent(self,i):
        return (i-1)//2
    
    def _left(self,i):
        return 2*i + 1
    
    def _right(self,i):
        return 2*i + 2

    def push(self,val):
        self.data.append(val)
        self._bubble_up(len(self.data)-1)
    
    def pop(self):
        if not self.data: #emtpy heap
            raise IndexError("pop from empty heap")
        smallest = self.data[0] #first item
        last = self.data.pop()

        if self.data: #if anyting is left, this needs to be reassigned 
            self.data[0] = last 
            self._bubble_down(0) #bubble down

        return smallest

    def peek(self):
        return self.data[0] #return min
    
    def __len__(self):
        return len(self.data)

    def _bubble_up(self,i):
        while i > 0:
            p = self.parent(i)
            if self.data[i] < self.data[p]: #if current is smaller,
            #then current needs to be swapped with aprent as current should be 
            # the root of the subtree, to stick to the rules of heaps. 
                self.data[i],self.data[p] = self.data[p], self.data[i]
                i = p 
            else:
                break

    def _bubble_down(self,i):
        n = len(self.data)
        while True: #tis 
            smallest = i     
            l = self._left(i)
            r  =self._right(i)
            #since we are bubbling down
            #swapping will need to consdier both left and right child nodes 
            # as opposed to just one 

            if l < n and self.data[l] < self.data[smallest]:
                smallest = l 
            if r < n and self.data[r] < self.data[smallest]:
                smallest = r
            if smallest == i:
                break
            self.data[i], self.data[smallest] = self.data[smallest], self.data[i]

            i = smallest #swap



#testing

if __name__ == '__main__':
    h = MinHeap()
    for x in [5, 3, 8, 1]:
        h.push(x)
        print("after push", x, "->", h.data)
 
    print("pop ->", h.pop(), " heap now", h.data)
 
    # self-check: popping everything must give the numbers in sorted order
    for _ in range(1000):
        nums = [random.randint(-50, 50) for _ in range(random.randint(0, 30))]
        mine = MinHeap()
        for x in nums:
            mine.push(x)
        assert [mine.pop() for _ in range(len(nums))] == sorted(nums)
    print("all random tests passed")

File: data_structures/data_structures_cpp/heaps.cpp
//heaps as vectors

//a heap is always a complete binry tree, meanin every level is full except perjaps the alst 
// push is bubble up, pop is bubble down 
//for bubble up, if it is smaller than its parent, keep swapping untilit isnt
//for bubble down you cannot delete index 0 since that would leave a hole. instead
//movethe last item into the root and then swap it downwards with its smaler chil until both children are bigger. it has to be the smaller chidl since if you swapped with the bigger one, that bigger child
//would become the parent of the smaller one, thus breaking the rule

//both push and pop operations are O(log n) and finding the top one is O(1)

#include <iostream>
#include <stdexcept>
#include <utility>
#include <vector>
using namespace std;

class MinHeap {
private:
    vector<int> data;   // the tree, stored level by level

    int parent(int i) { return (i - 1) / 2; }
    int left(int i)   { return 2 * i + 1; }
    int right(int i)  { return 2 * i + 2; }

    // New item at the bottom: swap it up while it's smaller than its parent.
    void siftUp(int i) {
        while (i > 0 && data[i] < data[parent(i)]) {
            swap(data[i], data[parent(i)]);
            i = parent(i);
        }
    }

    // Item at position i may be too big: swap it down with its smaller child.
    void siftDown(int i) {
        int n = data.size();
        while (true) {
            int smallest = i;
            int l = left(i);
            int r = right(i);
            if (l < n && data[l] < data[smallest]) smallest = l;
            if (r < n && data[r] < data[smallest]) smallest = r;
            if (smallest == i) break;    // both children are bigger: done
            swap(data[i], data[smallest]);
            i = smallest;
        }
    }

public:
    void push(int v) {
        data.push_back(v);               // add at the next free spot
        siftUp(data.size() - 1);         // restore the heap property
    }

    void pop() {
        if (empty()) throw out_of_range("pop on empty heap");
        data[0] = data.back();           // move the last item to the root
        data.pop_back();
        if (!empty()) siftDown(0);       // restore the heap property
    }

    int top() {
        if (empty()) throw out_of_range("top on empty heap");
        return data[0];                  // smallest item is always at the root
    }

    bool empty() { return data.empty(); }
    int size()   { return data.size(); }
};

int main() {
    MinHeap h;
    for (int x : {5, 3, 8, 1, 9, 2}) h.push(x);

    while (!h.empty()) {
        cout << h.top() << ' ';          // prints 1 2 3 5 8 9
        h.pop();
    }
    cout << '\n';
}